### PR TITLE
Implement crash management

### PR DESCRIPTION
**Pull request by tudorms**
### Type of Change
- New feature (non-breaking change which adds functionality)

### Why
We want to make runtime crashes more actionable by recording additional data with Windows Error Reporting which can be requested along crash reports and dumps for analysis.

### What / Why
* Add a CrashManager that calls into crashHandler functions exposed by the ReactInstances. Implement crashHandler for ReactInstance that forwards to JSIRuntimeHolder; the Hermes JSIRuntimeHolder calls into the Hermes VM crashHandler to record some additional details at the time of the crash (such as callstack and addresses of heaps)
* Replaced the Hermes runtime with one that registers additional details (memory blocks, metadata) with Windows Error Reporting to have them available in case of a crash
* Add a EnableDefaultCrashHandler property on the instance settings (default false) which -when enabled- registers a default unhandled exception handler which calls into the CrashManager to register a `CrashDetails.txt` file for WER (this is only uploaded by "requesting additional data" for the bucket in Windows Dev Center / Watson)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9150)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:19]: >g_wasCalled [](http://example.com/codeflow?start=5&length=11)

(nit) since this is in general Mso::React namespace, can we make naming more distinct, eg exceptionFilterWasCalled, or something like that

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:28]: > Cr [](http://example.com/codeflow?start=1&length=3)

(nit) should this be surrounded with try/catch or some other protection?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:23]: >EXCEPTION_EXECUTE_HANDLER [](http://example.com/codeflow?start=11&length=25)

This looks like a we are in really bad state if there is recursion. Could you add comment on the expected behavior if that happens?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.h
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.h:10]: >; [](http://example.com/codeflow?start=33&length=1)

(nit) noexcept?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:43]: >CrashDetails [](http://example.com/codeflow?start=23&length=12)

(nit) maybe ReactHostCrashDetails?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:33]: >RegisterCustomHandler [](http://example.com/codeflow?start=5&length=21)

Should we track whether this function can only be called once or multiple times?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:50]: >UnregisterCustomHandler [](http://example.com/codeflow?start=5&length=23)

Feels like this Register/Unregister should just have Raii semantics

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:59]: >) [](http://example.com/codeflow?start=51&length=1)

noexcept?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:68]: >fd [](http://example.com/codeflow?start=47&length=2)

(nit) spell out "fileDescriptor"?

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:194]: >s_mutex [](http://example.com/codeflow?start=40&length=7)

(nit) is this supposed to protect only s_entries? If so, can you name more specifically, eg s_entriesMutex

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (NikoAri) [reply vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:68]: Would be nice to rename in other .h files too

**Review by NikoAri — APPROVED** [review]
:shipit:

File: vnext/Microsoft.ReactNative/ReactApplication.cpp
Comment (asklar) [inline vnext/Microsoft.ReactNative/ReactApplication.cpp:45]: note that `UnhandledException` only gets triggered in _specific cases_ (@cglein can tell you all about how "easy" this is 😈). For example it won't be triggered if the exception happened in an async call in UWP... or maybe that was for win32?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (asklar) [reply vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:23]: an exception handler can, in turn, cause an exception :)
I think in some cases, double faults should just kill the app, it's unlikely we'll recover gracefully enough

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (asklar) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:35]: adding a comment here so we figure out the story for official apps (uwp) before merging. I think you should be able to use the CORE_API define or one of the helpers to find out if it is a UWP app

File: vnext/Microsoft.ReactNative/ReactInstanceSettings.idl
Comment (asklar) [inline vnext/Microsoft.ReactNative/ReactInstanceSettings.idl:155]: If it is default, then it should be turned on? Otherwise it isn't the default :)

File: vnext/Shared/HermesShim.cpp
Comment (asklar) [inline vnext/Shared/HermesShim.cpp:78]: I think this means apps that were created with let's say RNW 0.66, would now crash when upgraded to 67 because the export is not found in _their_ version of Hermes
Can we fail gracefully instead?

File: vnext/Shared/HermesShim.cpp
Comment (asklar) [inline vnext/Shared/HermesShim.cpp:15]: can't see the signature here but can we guarantee that this `hermesCrashHandler` function is noexcept / doesn't throw since this is across DLL boundaries (plus it should throw in a WER handler)

**Comment by asklar** [comment]
Note that we already have a terminate handler when calling out to modules: see `TerminateExceptionGuard`:

https://github.com/microsoft/react-native-windows/blob/main/vnext/Microsoft.ReactNative/IReactModuleBuilder.cpp#L23-L65

we should rationalize how this plays with that

**Comment by asklar** [comment]
I'd like to understand better how I'd be able to access the data. As it is it seems like this asks Hermes to save some info to a local file called CrashDetails.txt. I'd have to manually go into Watson and request that we start collecting this data, so this won't work for broad usage for many of our customers (who will be debugging crash dumps from Watson).

How big is this file? Is it possible to instead log to memory and make sure WER includes that piece of data when uploading?

File: vnext/Microsoft.ReactNative/ReactInstanceSettings.idl
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactInstanceSettings.idl:155]: A naming issue; maybe "baseline" or "reference implementation / built-in" instead of default?

File: vnext/Shared/HermesShim.cpp
Comment (tudorms) [reply vnext/Shared/HermesShim.cpp:78]: This whole shim thing feels weird, not sure what it's for exactly. But I suspect apps would need to version match RNW and Hermes, right? We don't support N-1 / N+1 with random Hermes binaries.

**Comment by tudorms** [comment]
By default WER only sends a report (or a mini-dump, depending on the machine's privacy settings). The app owner would need to go to Windows Dev Center externally (or Watson internally) to "request more data" which will trigger WER sending heap dumps that include the registered memory blocks and the CrashDetails.txt file. There is no way to get CrashDetails.txt uploaded except by requesting more data (confirmed with wercore), unless we implement our own pipeline.

If developers don't have a Windows Dev Center account (previously called Winqual) I don't think there's any way to get this data, unless they instruct their app's customers to manually retrieve it from their LocalFolder path (something like %APPDATA%\Local\Packages\RNPlayground_8wekyb3d8bbwe\LocalState\CrashDetails.txt).

WER does offer APIs for manually creating a report / minidump, so I guess we can implement a parallel pipeline where we call the minidump APIs instead (WerReportCreate, WerReportAddDump) which again the developers could instruct their customers to manually upload with support requests (probably an edge case, which we could explore later, depending on customer requests).

For the other question, WER custom metadata is limited to 128 ASCII characters, and there is no programmatic way to include certain memory blocks in a mini-dump (we can only include memory blocks in heap or larger dumps, see [https://docs.microsoft.com/en-us/windows/win32/api/werapi/nf-werapi-werregistermemoryblock](https://docs.microsoft.com/en-us/windows/win32/api/werapi/nf-werapi-werregistermemoryblock) ).

---
In reply to: [971044346](https://github.com/microsoft/react-native-windows/pull/9150#issuecomment-971044346) [](http://example.com/codeflow?ancestors=971044346,971044346)

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h:195]: >s_entries [](http://example.com/codeflow?start=41&length=9)

nit: it would be nice to have a more descriptive name. E.g. s_instanceRegistry or something else that says that these are all running instances.

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h:85]: >fd [](http://example.com/codeflow?start=31&length=2)

nit: it would be nice to have a more descriptive parameter name. Is it a file descriptor?

**Comment by tudorms** [comment]
Looks like that is debug-only so it's probably not very interesting. Also unclear when it would be triggered.

FWIW, I tested the code in this PR with an access violation in debug builds and that handler wasn't called (the UnhandledExceptionHandler set in CrashManager.cpp was called). We can try with a few other exception types I guess.

---
In reply to: [971037730](https://github.com/microsoft/react-native-windows/pull/9150#issuecomment-971037730) [](http://example.com/codeflow?ancestors=971037730,971037730)

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:13]: nit: The comment is a little bit confusing to read. Does it miss upper case letters to start sentenses and dots to finish them?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:16]: >g_previousExceptionFilter [](http://example.com/codeflow?start=29&length=25)

It is better to use the `static` keyword for the variables specific only to this cpp file. It helps to avoid any unexpected linker conflicts.

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:43]: Should it be app-specific?

File: vnext/Microsoft.ReactNative/ReactInstanceSettings.idl
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/ReactInstanceSettings.idl:155]: I wonder if should reverse it and have the DisableDefaultCrashHandler. I guess the main questions is if this default allowes third party developers to get the crash reports. If yes, then it should be 'on' by default.

File: change/react-native-windows-37128f50-e855-4562-9019-16b5480fceb4.json
Comment (vmoroz) [inline change/react-native-windows-37128f50-e855-4562-9019-16b5480fceb4.json:3]: >CrashManager [](http://example.com/codeflow?start=14&length=12)

nit: it would be better to expand it.

Comment (acoates-ms) [reply change/react-native-windows-37128f50-e855-4562-9019-16b5480fceb4.json:3]: Agreed, - this string shows up in the changelog

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:37]: Pretty sure that for things like office, this would be a directory that the app wouldn't have write access to.  Should we be using the temp dir instead?

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:251]: This looks like a global state, rather than anything instance specific.  -- Why is this tried to a ReactInstance Settings, rather than just being a static function that should be called at a per app level?

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:254]: Isn't this the bit that should be controlled by the instance settings?

File: vnext/Microsoft.ReactNative/ReactApplication.cpp
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactApplication.cpp:52]: ReactApplication only handles the most basic type of RNW app.  Everything ReactApplication does should be accessible to more complex apps that are not using ReactAppliction.  So we should add some kind of OnUnhandledException function to one of the IDLs.

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:251]: Do we have existing examples of that? Is it common for apps to have more than one ReactNativeHost / ReactInstance?

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:28]: I marked OnUnhandledException as noexcept just in case.

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:43]: Called it `ReactNativeCrashDetails.txt` for clarity.

File: vnext/Microsoft.ReactNative/ReactApplication.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactApplication.cpp:45]: I found very little and conflicting information about SetUnhandledExceptionFilter vs. CoreApplication::UnhandledException . I suspect SetUnhandledExceptionFilter actually takes care of most native exception types; however, if there is managed code in the application I've found comments online saying unhandled exceptions in managed code don't actually trigger SetUnhandledExceptionFilter . but they do trigger CoreApplication::UnhandledException.
So really this is mostly optional for catching a few additional types of exceptions and routing them to the handler.

Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactApplication.cpp:45]: [https://www.xsprogram.com/content/how-does-setunhandledexceptionfilter-work-in-net-winforms-applications.html](https://www.xsprogram.com/content/how-does-setunhandledexceptionfilter-work-in-net-winforms-applications.html)

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp
Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:251]: Its not common, but Office for one.

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:254]: I think there is value in tracking all active instances somewhere, but I added a check for the opt-in setting in InstanceCrashHandler.

**Comment by NikoAri** [comment]
Just to clarify from previous email thread, Watson will collect some of its Process Dumps as Heap Dumps automatically.

---
In reply to: [971091655](https://github.com/microsoft/react-native-windows/pull/9150#issuecomment-971091655) [](http://example.com/codeflow?ancestors=971091655,971044346)

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:251]: Figured out a cleaner design where we inc/dec for each react instance that opts in. That will cover cases where apps have multiple instances with differing options.

File: vnext/Microsoft.ReactNative/ReactApplication.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactApplication.cpp:52]: Do IDLs support static functions? Or do we have a static class somewhere for this purpose? I agree we need to expose it, but not sure where it belongs.

File: vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactHost/CrashManager.cpp:23]: Based on the documentation ([https://docs.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-unhandledexceptionfilter](https://docs.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-unhandledexceptionfilter)) it looks like this is the only value we can return here to avoid recursing. They suggest returning to allow the exception handler to do its thing; the alternative is to terminate the process, yes, though not sure if that'll allow WER to finish its collecting.

File: vnext/Shared/HermesShim.cpp
Comment (tudorms) [reply vnext/Shared/HermesShim.cpp:15]: It's not marked noexcept, but HermesRuntimeHolder::crashHandler (the wrapper) is.

I think we'll address the DLL boundary issue further with the ABI safety work.

File: vnext/Microsoft.ReactNative/ReactInstanceSettings.idl
Comment (tudorms) [reply vnext/Microsoft.ReactNative/ReactInstanceSettings.idl:155]: I'm not sure about this, the change could look invasive. I recommend starting with opt-in and seeing if we get any feedback on it.

File: vnext/Shared/HermesShim.cpp
Comment (NickGerleman) [reply vnext/Shared/HermesShim.cpp:78]: It's to be kosher with WACK

WACK tries to prevent you from shipping with missing DLLs, and will fail validation if you link to a function exported by a DLL not present. This is still safe when the DLL is delay-loaded, but WACK does not know about it.

We wanted a single Microsoft.ReactNative DLL, but the ability to opt-out of packaging Hermes DLL. This lead to loading the DLL at runtime, from its publicly exported function names (the mangled ones).

**Pull request closed without merging**
